RFCs/FS-1079-union-properties-visible.md

```diff
@@ -0,0 +1,118 @@
+    | _ -> false
+```
+
+Considering that F# already generates a member `IsEmail : bool` which does exactly the same as the above, it would make sense to make it available to F# developers so that the example becomes:
+
+```fsharp
+let canSendEmailTo person =
+    person.contact.IsEmail
+```
+
+# Detailed design
+
+The compiler generates the following public members for every discriminated union type:
+
+1. A property per case named `IsFoo : bool` for a case named `Foo`. This property is true if `this` is an instance of `Foo` and false otherwise.
+
+2. A property `Tag : int` whose value identifies the case: it is `0` if `this` is an instance of the first case (in type declaration order), `1` if `this` is an instance of the second case, and so on.
+
+3. A nested static type `Tags`.
+
+4. Inside `Tags`, a constant integer static field per case, with the same name as the case. Its value is the same as the `Tag` property of an instance of the corresponding case.
+
+5. Either a static property `Foo` or a static method `NewFoo` per case named `Foo`, depending on whether `Foo` has arguments or not, which constructs an instance of this case.
+
+In this design, members #1 through 4 are made available to F# code as well. #5 is kept hidden from F#, because it is equivalent to the corresponding case constructor and would be redundant and confusing.
+
+Here is an example union type declaration with all of the newly F#-visible members listed explicitly (in pseudo-code, since nested `type` declarations are invalid F#):
+
+```fsharp
+type Example =
+    | First
+    | Second of int * string
+    
+    // The following members were already generated, and are now visible to F#:
+    
+    member IsFirst : bool
+    
+    member IsSecond : bool
+    
+    member Tag : int
+    
+    type Tags =
+    
+        val First : int
+        
+        val Second : int
+        
+    // The following members were already generated, and remain hidden from F#:
+    
+    static member First : Example
+    
+    static member NewSecond : int * string -> Example
+```
+
+# Drawbacks
+
+Exposing `Tag` and `Tags` changes whether reordering the cases of a union type declaration is a breaking change:
+
+|                                     | Before this change   | After this change   |
+| :---------------------------------- | :------------------: | :-----------------: |
+| Binary compatible for F# consumer   | No                   | No                  |
+| Source compatible for F# consumer   | Yes                  | No                  |
+| Binary compatible for C# consumer   | No                   | No                  |
+| Source compatible for C# consumer   | No                   | No                  |
+
+# Alternatives
+
+An alternate design would be to only expose the `Is*` properties but not the `Tag` property and the `Tags` nested type. This would eliminate the drawback regarding case reordering, while still exposing what is arguably the most useful generated members.
```

Comment:
I generally agree with this perspective. I'd like to see a compelling use case for the use of Tag. I think the best one I can think of is efficient serialization, maybe? You could match on Tag (an int) instead of doing repeated `IsX` calls to find the case in question...

Reply:
Come to think of it, I actually agree too. I'll swap it around, make the RFC about just IsX and put adding Tag in the Alternatives section.